File: src/main/java/net/codinux/kotlin/Collections.java
package net.codinux.kotlin;

import java.util.List;
import java.util.stream.Collectors;

public class Collections {

    public long count(List<Long> clientIds) {
        return clientIds.stream().filter(id -> id > 10).count();
    }

    public List<String> map(List<Long> clientIds) {
        return clientIds.stream().map(id -> id.toString()).collect(Collectors.toList());
    }

    public List<String> mapNotNull(List<Long> clientIds) {
        return clientIds.stream().filter(id -> id != null).map(id -> id.toString()).collect(Collectors.toList());
    }

    public void joinToString(List<Long> clientIds) {
        System.out.println(clientIds.stream().map(id -> id.toString()).collect(Collectors.joining(", ")));
    }

    public void joinToStringWithOptions(List<Long> clientIds) {
        System.out.println(clientIds.stream().map(id -> "ID: " + id).collect(Collectors.joining("; ", "clientIds: ", ".")));
    }

    public List<Long> instanceOf(List<Number> clientIds) {
        return clientIds.stream().filter(id -> id instanceof Long).map(id -> (Long) id).collect(Collectors.toList());
    }

}

File: src/main/kotlin/net/codinux/kotlin/Collections.kt
package net.codinux.kotlin

import net.codinux.kotlin.model.Person
import java.time.LocalDate


class Collections {

  fun count(clientIds: List<Long>): Int {
    return clientIds.count { it > 10 }
  }


  fun map(clientIds: List<Long>): List<String> {
    return clientIds.map { it.toString() }
  }

  fun mapNotNull(clientIds: List<Long>): List<String> {
    return clientIds.mapNotNull { it.toString() }
  }


  fun joinToString(clientIds: List<Long>) {
    println(clientIds.joinToString()) // prints e.g. "7, 19, 33, 56"
  }

  fun joinToStringWithOptions(clientIds: List<Long>) {
    println(clientIds.joinToString("; ", "clientIds: ", ".") { "ID: $it" }) // e.g. "clientIds: ID: 7; ID: 19; ID: 33."
  }


  fun instanceOf(clientIds: List<Number>): List<Long> {
    return clientIds.filterIsInstance<Long>()
  }


  fun groupBy(persons: List<Person>): Map<LocalDate, List<Person>> {
    return persons.groupBy { it.dayOfBirth }
  }

  fun toMap(persons: List<Person>): Map<String, Person> {
    return persons.associateBy { it.name }
  }

  fun toMapMapped(persons: List<Person>): Map<LocalDate, String> {
    return persons.associate { it.dayOfBirth to it.name }
  }

}